### DISCUSSION TITLE
NEXT.JS Fetch api with env variable as key when button clicked

### DISCUSSION THREAD

**Discussion by Ronuhz** [vercel/next.js#52055] [Help] [answered]
Next.js version: 13.4

I have a component which returns a some elements among those a button. I also have a function like
```
async function getDataFromApi(name) {
  const res = fetch(`http://dummy.api?name=${name}`,
    {
      headers: {
        Authorization: process.env.API_KEY,
      },
    }
  )
  const data = await res.json()
  return data
}
```

How can I run this function whenever the button gets clicked? I tried using both client (can’t because of the process.env.API_KEY would return undefined ) and server (can’t because of the onClick() on the button and the full code uses states as well) components.
Also the API key shouldn’t be exposed, because it wouldn't be great if my key gets leaked.

**Comment by rasif-sahl** [comment, marked as answer]
To run the getDataFromApi function when a button is clicked in Next.js while also ensuring the API key is not exposed, you can make use of an API route and handle the API call server-side.

Here's how you can accomplish this:

1. Create an API route in your Next.js project. Create a file called my-api.js in the pages/api directory:

```
// pages/api/my-api.js

export default async function handler(req, res) {
  const { name } = req.query;

  // Perform API call here using the name parameter and the API key

  try {
    const apiResponse = await getDataFromApi(name);
    res.status(200).json(apiResponse);
  } catch (error) {
    res.status(500).json({ error: 'An error occurred' });
  }
}

```
In this example, the API route receives a name parameter from the query string. It calls the getDataFromApi function and returns the API response as JSON.

2. Modify your component to use the API route and handle the button click:

```
import { useState } from 'react';

function MyComponent() {
  const [data, setData] = useState(null);

  const handleClick = async () => {
    try {
      const response = await fetch(`/api/my-api?name=yourName`);
      const apiResponse = await response.json();
      setData(apiResponse);
    } catch (error) {
      console.error('Error:', error);
    }
  };

  return (
    <div>
      <button onClick={handleClick}>Fetch Data</button>
      {data && <div>{/* Render the data */}</div>}
    </div>
  );
}

export default MyComponent;

```

In this code, the handleClick function is called when the button is clicked. It makes a request to the API route /api/my-api with the desired name parameter. The API route calls the getDataFromApi function and returns the response, which is then set as the component's state.

By using an API route, you can handle the API call server-side, protecting your API key from exposure on the client-side. The client only interacts with the API route, and the server handles the actual API call using the key from the server environment.

Remember to replace 'yourName' in the handleClick function with the desired value for the name parameter.

This approach ensures that the API key remains secure while allowing you to fetch data from the API when the button is clicked in your Next.js component.

**Reply by Ronuhz** [reply]
Thank you very much.
